client/packages/programs/src/JsonForms/components/ProgramEvent.tsx

```diff
@@ -0,0 +1,228 @@
+ */
+const Options = z
+  .object({
+    /**
+     * The time when the event should be triggered.
+     * Time of the event:
+     * `before`: just before the current encounter
+     * `after`: at and after the start of the current encounter
+     * Default: `before`
+     */
+    at: z
+      .discriminatedUnion('type', [
+        z.object({
+          type: z.literal('encounter'),
+          encounterStartDatetime: z.enum(['before', 'after']).optional(),
+        }),
+        z.object({
+          type: z.literal('programEnrolment'),
+          programEnrolmentDatetime: z.enum(['before', 'after']).optional(),
+        }),
+      ])
+      .optional(),
+    documentType: z.string(),
+    eventType: z.string(),
+    /**
+     * Display option based on type.
+     */
+    display: z
+      .discriminatedUnion('type', [
+        z.object({ type: z.literal('number') }),
+        z.object({
+          type: z.literal('string'),
+          multiline: z.boolean().optional(),
+          rows: z.number().optional(),
+          show: z.array(
+            z
+              .tuple([z.string(), z.string().optional()])
+              .rest(z.string().optional())
+          ),
+        }),
+      ])
+      .optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+const extractAt = (datetime?: string, options?: Options): Date => {
+  if (!datetime) {
+    return new Date();
+  }
+  const beforeDate = new Date(new Date(datetime).getTime() - 1);
+  if (!options || !options.at) {
+    return beforeDate;
+  }
+
+  if (options.at.type === 'encounter') {
+    switch (options.at.encounterStartDatetime) {
+      case 'before':
+        return beforeDate;
+      case 'after':
+        return new Date(datetime);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.encounterStartDatetime as never);
+    }
+  } else if (options.at.type === 'programEnrolment') {
+    switch (options.at.programEnrolmentDatetime) {
+      case 'before':
+        return beforeDate;
+      case 'after':
+        return new Date(datetime);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.programEnrolmentDatetime as never);
+    }
+  }
+
+  return new Date();
+};
+
+const getDisplayOptions = (
+  data: string | null | undefined,
+  options?: Options
+) => {
+  const show =
+    options?.display?.type === 'string' ? options?.display?.show : null;
+
+  if (!show) {
+    return data;
+  }
+
+  const displayValue = show.find(value => value[0] === data)?.[1];
+  return displayValue ?? '';
+};
+
+const UIComponent = (props: ControlProps) => {
+  const { label, uischema, config } = props;
+  const [datetime, setDatetime] = React.useState<string | undefined>();
+  const patientId = config?.patientId;
+  const encounterDocName = config?.documentName.includes('Encounter');
```

Comment:
Checking the document name and passing it onto the hook so that both hooks don't get called at the same time (which can result in an error)

Reply:
see my other comment how this can be solved in a "cleaner" way without making assumptions on the document name...

Reply:
please use the validate zod config that has the correct type...